File: leetcode/algorithms/c++/Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.cpp
class Solution
{
public:
    int tab[255];

    int lengthOfLongestSubstring(string s)
    {
        if (s.size() == 0)
            return 0;
        if (s.size() == 1)
            return 1;
        for (int i = 0; i < 255; i++)
            tab[i] = 0;

        int result = 1;
        int begin = 0;
        int end = 1;
        ++tab[s[begin]];
        while (begin < s.size())
        {
            if (begin >= end)
            {
                end = begin + 1;
                for (int i = 0; i < 255; i++)
                    tab[i] = 0;
                ++tab[s[begin]];
            }
            while (end < s.size() && check(s[end]))
            {

                ++tab[s[end]];
                ++end;
            }

            int val = end - begin;
            if (val > result)
                result = val;

            --tab[s[begin]];

            ++begin;
        }

        return result;
    }

    bool check(char x)
    {
        if (tab[x] > 0)
            return false;
        return true;
    }
};

File: leetcode/algorithms/python/Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.py
class Solution:
    tab = {}

    def check(self, s: chr) -> bool:
        if (s not in self.tab):
            return True
        if (self.tab[s] > 0):
            return False
        return True

    def lengthOfLongestSubstring(self, s: str) -> int:
        if (len(s) == 0):
            return 0
        if (len(s) == 1):
            return 1

        result = 1
        begin = 0
        end = 1

        self.tab[s[begin]] = 1
        while(begin < len(s)):
            if (begin >= end):
                end = begin + 1
                self.tab = {}
                self.tab[s[begin]] = 1
            while(end < len(s) and self.check(s[end])):
                if s[end] in self.tab:
                    self.tab[s[end]] += 1
                else:
                    self.tab[s[end]] = 1
                end += 1
            val = end - begin
            if (val > result):
                result = val
            self.tab[s[begin]] -= 1
            begin += 1
        return result
